### ISSUE TITLE
Nordic characters (äöå) in filenames cannot be pushed to Mercurial repository with SCM hooks

### ISSUE THREAD

**Issue by Hannu1** [scm-manager/scm-manager#1911]
SCM Manager (2.27.4) hooks throw an exception while pushing a file with filename that contains non ascii characters like nordic characters (äöå). This aborts the push operation. Nordic characters work fine with hg without scm.

Reproduce:
1. create repository in scm
2. clone locally
2. commit file with name äbc.txt
3. try to push

Exception:
ERROR onia.scm.repository.spi.hgHookChangesetProvider - could not retrieve changesets
com aragost.javahg.internals.RuntimeIOException: Input length = 1
   at com aragost.javahg.internals.Utils.asRuntime(Utils.java:437)
   at com aragost.javahg.internals.Utils.decodeBytes(Utils.java:243)
   at com aragost.javahg.internals.Utils.decodeBytes(Utils.java:223)
   at com aragost.javahg.internals.HgInputStream.textUpTo(HgInputStream.java:430)
   at sonia.scm.repository.spi.javahg.AbstractChangesetCommand.createFromInputStream(AbstractChangesetCommand.java:240)
   at sonia.scm.repository.spi.javahg.AbstractChangesetCommand.readListFromStream(AbstractChangesetCommand.java:164)
   at sonia.scm.repository.spi.javahg.HgLogChangesetCommand.execute(HgLogChangesetCommand.java:65)
   ...
Caused by: java.nio.charset.MalformedInputException: Input length = 1
   at java.base/java.nio.charset.CoderResult.throwException(CoderResult.java:274)
   at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:813)
   at com.aragost.javahg.internals.Utils.decodeBytes(Utils.java:241)
   ...

EDIT:
SCM Manager installed on RHEL server, hg version is 4.8.2
Clients are on Windows, mostly TortoiseHG and HG version 5.9.1.

**Comment by pfeuffer** [comment]
I tried to reproduce this with a Fedora docker image, but could not found any issues. Do you have configured a character set other than UTF-8 on your server? Maybe you can install the [Script Plugin](https://scm-manager.org/plugins/scm-script-plugin/) and run the following little code to determine your charset:

```
println java.nio.charset.Charset.defaultCharset()
```

**Comment by Hannu1** [comment]
It's rather complicated process to install plugins, since our server is running in an offline environment.

I tried echo $LANG from console and got "en_US.UTF-8"
I also ran a small java program that printed java.nio.charset.Charset.defaultCharset() and got "UTF-8".

If there's a scm manager specific configuration for this, I can check that value as well.

I ran some more tests as well. The files with nordic characters have been added to the repository from Windows  workstation. If I create, add, commit and push a file (äbc.txt) from Linux, the push succeeds. If I pull to a Windows machine after this, the characters are scrambled, which makes sense with encoding conflict.

**Comment by Hannu1** [comment]
We can stop using non-ascii characters in the future. The issue is that we imported an old repository (with 30k revs) that worked with an old version of SCM Manager. I can't get the exact version right now, because the server has already been shutdown and I can't start it. I think it was 21. 

Now we cannot rename the files in the old repository because it won't let us push a commit that modifies the old files in anyway. Commits to some older branches cannot be pushed, because fetching changesets crashes when it finds the files with nordic characters.

**Comment by pfeuffer** [comment]
OK, now I got it. You can configure Mercurial to use CP-1252 for instance. Just set "Encoding" in the main configuration for Mercurial to `cp1252` in SCM-Manager. Just be sure, that you only use Windows clients from then on.

**Comment by Hannu1** [comment]
Thank you! Now pushes to that repository work again. Do you know, if we remove all non-ascii characters from filenames, is it possible to switch back to utf-8?

Any plans on making encoding a "per repository" setting? Would like to have our other repositories use UTF-8. No issues arose with cp1252 though.

**Comment by eheimbuch** [comment]
Hey @Hannu1,

we have currently no plans about such a setting on repository level. But if you need it feel free to implement it yourself. We could support you if you have concrete questions regarding the implementation.

**Comment by Hannu1** [comment]
I looked into implementing this today. Looking at the code this was already an option in HgRepositoryConfig with everything seemingly ready on the java side. Figured I'd add it to the webapp and went to look for a place for it on the settings page of the repository and found the option already there as well. My bad :) Somehow I'd missed it before. Sorry for the confusion.

So, if someone stumbles to this issue: 
1. There's the global option for hg encoding in "Administration > Settings > Mercurial" page 
2. A repository specific option can be found on that repository's settings page. Tested it and it overrides global option as expected.
